Final/SyncLinkedList.h

```diff
@@ -0,0 +1,31 @@
+#ifndef SYNC_LINKED_LIST_H
+#define SYNC_LINKED_LIST_H
+
+#include <shared_mutex>
+#include <mutex>
+#include <functional>
+#include "LinkedList.h"
+
+class LinkedList;
+
+class SyncLinkedList {
+public:
+    // Constructor
+    SyncLinkedList();
+    // Destructor
+    ~SyncLinkedList();
+
+    // Pushes node to front of list
+    void push(int data);
+    // Removes the first instance of a node with the given data
+    void remove(int data);
+
+    // Exposed to allow for easy iteration through the linked list
+    // If the iter handler returns false, the iteration will stop.
+    void for_each(std::function<bool (node *)> iter);
+private:
+    LinkedList *linked_list;
```

Comment:
i keep forgetting we're in c++ world. in reality, we can just make this a non pointer (i.e. `LinkedList linked_list`), then we don't need to deal with allocating/freeing things in the constructor/destructor. just by declaring the variable, the underlying object will be created and initialized using the default constructor for the class, but it will live on the stack instead of the heap meaning that its memory allocation/deallocation will be automatically handled by the operating system instead of us needing to manually free things.

Note that this is different from how Java handles objects, where you _always_ need to `new` things. for example, in java:
```
ArrayList<Integer> list = new ArrayList<Integer>();
list.add(1); // This is okay
ArrayList<Integer> list2;
list2.add(1); // This will crash
```
But in C++ world:
```
vector<int> *list = new vector<int>();
list->push_back(1); // This is okay - note that we manually allocated a vector on the heap using new, resulting in us being returned a pointer to a vector (vector<int> *)
vector<int> list;
list.push_back(1); // This is okay as well - note how this is note a pointer and so we can use . instead of ->
```